src/sorta.nim

```diff
@@ -409,45 +421,83 @@ proc `$`*[Key, Val](b: SortedTable[Key, Val]): string =
 
 proc len*[Key, Val](b: SortedTable[Key, Val]): int {.inline.} = b.entries
 
-
-iterator entries[Key, Val](b: SortedTable[Key, Val]): Entry[Key, Val] =
-  template populateWithLeftmostChildren(stack, currentNode) =
-    while not currentNode.isNil:
-      for i in countdown(currentNode.m-1, 0):
-        stack.add(currentNode.e[i])
-      currentNode = currentNode.p0
-
-  var
-    root = b.root
-    currentEntry: Entry[Key, Val]
-    stack = newSeq[Entry[Key, Val]]()
-
-  stack.populateWithLeftmostChildren(root)
-  while stack.len > 0:
-    currentEntry = stack.pop()
-    yield currentEntry
-    stack.populateWithLeftmostChildren(currentEntry.p)
-
+proc key[Key, Val](cursor: CursorPosition[Key, Val]): Key =
+  ## Return the key for a given cursor position
+  cursor.node.e[cursor.entry].key
+
+proc val[Key, Val](cursor: CursorPosition[Key, Val]): Val =
+  ## Return the value for a given cursor position
+  cursor.node.e[cursor.entry].val
+
+proc search[Key, Val](b: SortedTable[Key, Val], key: Key): Cursor[Key, Val] =
+  ## calculate the cursor pointing to the given key
+  var a = b.root
+  while not a.isNil:
+    var r = binarySearch(key, a)
+    if r < a.m:
+      result.stack.add((a, r))
+      if eq(key, a.e[r].key):
+        break
+    a = if r == 0: a.p0 else: a.e[r-1].p
+  # add a dummy entry for first next call
+  result.stack.add((nil, 0))
+
+proc current[Key, Val](cursor: Cursor[Key, Val]): CursorPosition[Key, Val] =
+  ## the current position of a cursor
+  last(cursor.stack)
+
+proc mcurrent[Key, Val](cursor: var Cursor[Key, Val]): var CursorPosition[Key, Val] =
+  ## a reference to the current position of a cursor
+  mlast(cursor.stack)
+
+proc next[Key, Val](cursor: var Cursor[Key, Val]): bool =
+  ## move cursor forward returning true if cursor.current is now valid.
+  ## never call next after it returns false.
```

Comment:
Style: all doc comments (lines starting with `##`) should be written in "sentence case": first character is upper case, and it ends with the dot; also: it explains what the proc does so it is "moves ....; calls ...." instead of "move ....; call ..."

E.g.:
```suggestion
  ## Moves cursor forward returning true if cursor.current is now valid.
  ## Never calls next after it returns false.
```

This stuff should be changed in other places too.

Reply:
Fixed